### ISSUE TITLE
Question: is proxy change possible without restarting application?

### ISSUE THREAD

**Issue by notreallymyself** [rofl0r/proxychains-ng#399]
I am wrapping proxychains-ng around a long-living process (mitmproxy) in a scenario where I need to change a single outgoing proxy (defined in the proxychains.conf file) without restarting my application and losing its state (flow history, preferences, event log, etc).

proxychains.conf:
```
random_chain
chain_len = 1
quiet_mode
proxy_dns
tcp_read_time_out 10000
tcp_connect_time_out 8000
[ProxyList]
socks5   x.x.x.x xxxx
```

launching mitmproxy wrapped around proxychains-ng:
```
proxychains4 -f ~/p.conf mitmproxy --listen-port xxxx ... etc
```

Let's say I'm running mitmproxy, and the proxy defined in the config dies, if I want to change it to an online proxy, I have to edit the config file, close mitmproxy, and relaunch it so the change is applied, it works, but I lose all the information/configurations in mitmproxy.

I know I can define a list of proxies and proxychains will skip dead ones automatically, or use a random one from the list, and so on, but in my case I require a single proxy and I need to control it manually very closely, rather than letting proxychains do the automatic/random proxy switching.

**In short, I basically need to switch an outgoing proxy to a specific proxy of my choice while my program is running, without restarting my application and losing its state**, is such a thing even possible? 

I know proxychains-ng doesn't monitor the config file supplied in the parameters at intervals and correct its behavior based on the changes to the config file, because I tested for that, do I have any options here?

P.S. I love this project, it has saved me a ton of headaches :)

**Comment by notreallymyself** [comment]
I came up with a possible solution, perhaps I can combine proxychains-ng and tinyproxy to avoid restarting mitmproxy's process every time I need to change the proxy defined in proxychain's config file.

instead of doing this:
`mitmproxy -> proxychains -> remote upstream proxy`
I could set it up like this:
`mitmproxy -> proxychains -> local tinyproxy instance -> remote upstream proxy`

With this setup, I could modify the configured upstream proxy in tinyproxy, and restart tinyproxy every time I need to switch proxies around, avoiding the need to restart mitmproxy/proxychains directly.

I'm still curious to know if there's other 'cleaner' solutions, though.

**Comment by rofl0r** [comment]
no, proxychains doesn't support this directly, but the tinyproxy solution you came up with is nice. actually tinyproxy config can be changed at runtime by signaling the running process with SIGUSR1, so you don't have to restart it.

**Comment by notreallymyself** [comment]
Thanks for the SIGUSR1 tip, I am testing this setup right now and I can confirm that it works perfectly to solve all the problems I was having. Cheers!
